util/HalideTraceDump.cpp

```diff
@@ -0,0 +1,391 @@
+    }
+
+    printf("Done.\n");
+}
+
+void usage(char * const *argv) {
+    const string usage =
+            "Usage: " + string(argv[0]) + " -i trace_file -t output_type\n"
+            "  Available output types: png, jpg, pgm, tmp\n"
+            "The tool will record all element loads and stores found in the trace, and "
+            "dump them into separate image files per Func.\n"
+            "To generate a trace, use the trace_loads() and trace_stores() scheduling "
+            "commands in your Halide application, and run with HL_TRACE_FILE=<filename>.\n"
+            "Only 2D buffers will be dumped.";
+    fprintf(stderr, "%s\n", usage.c_str());
+    exit(1);
+}
+
+int main(int argc, char * const *argv) {
+
+    char *buf_filename = nullptr;
+    char *buf_imagetype = nullptr;
+    BufferOutputOpts outputopts;
+    int c;
+    while ((c = getopt (argc, argv, "i:t:")) != -1) {
+        switch (c)
+        {
+        case 'i':
+            buf_filename = optarg;
+            break;
+        case 't':
+            buf_imagetype = optarg;
+            break;
+        case '?':
+        default:
+            usage(argv);
+        }
+    }
+
+    if (buf_filename == nullptr) {
+        usage(argv);
+    }
+    if (buf_imagetype == nullptr) {
+        usage(argv);
+    }
+
+    string imagetype(buf_imagetype);
+    if (imagetype == "jpg") {
+        outputopts.type = BufferOutputOpts::JPG;
+    } else if (imagetype == "png") {
+        outputopts.type = BufferOutputOpts::PNG;
+    } else if (imagetype == "pgm") {
+        outputopts.type = BufferOutputOpts::PGM;
+    } else if (imagetype == "tmp") {
+        outputopts.type = BufferOutputOpts::TMP;
+    } else {
+        usage(argv);
+    }
+
+    FILE *file_desc = fopen(buf_filename, "r");
+    if (file_desc == nullptr) {
+        fprintf(stderr, "[Error opening file: %s. Exiting.\n", argv[1]);
+        exit(1);
+    }
+
+
+    printf("[INFO] Starting parse of binary trace...\n");
+    int packet_count = 0;
+
+    map<string, FuncInfo> func_info;
+
+    printf("[INFO] First pass...\n");
+
+    for (;;) {
+        Packet p;
+        if (!p.read_from_filedesc(file_desc)) {
+            printf("[INFO] Finished pass 1 after %d packets.\n", packet_count);
+            break;
+        }
+
+        // Packet read was successful.
+        packet_count++;
+        if ((packet_count % 100000) == 0) {
+            printf("[INFO] Pass 1: Read %d packets so far.\n", packet_count);
+        }
+
+        // Check if this was a store packet.
+        if ( (p.event == halide_trace_store) || (p.event == halide_trace_load) ) {
+            if (func_info.find(string(p.func())) == func_info.end()) {
+                printf("[INFO] Found Func with tracked accesses: %s\n", p.func());
+                func_info[string(p.func())] = FuncInfo(&p);
+            }
+            func_info[string(p.func())].add_preprocess(&p);
+        }
+    }
+
+    packet_count = 0;
+    fseek(file_desc, 0, SEEK_SET);
```

Comment:
do we also need to check return value of fseek?

Reply:
I don't think so, looks like it checks ferror immediately below, which will be set if fseek failed